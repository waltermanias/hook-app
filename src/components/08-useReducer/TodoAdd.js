import React, { useRef } from "react";
import { useForm } from "../../hooks/useForm";

export const TodoAdd = ({ handleAddTodo }) => {
  const { formValues, handleInputChange, reset } = useForm({ description: "" });

  const input = useRef();

  const handleSubmit = (e) => {
    e.preventDefault();
    if (formValues.description.trim() === "") {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: formValues.description,
      done: false,
    };

    handleAddTodo(newTodo);

    reset();
    input.current.select();
  };

  return (
    <>
      <h4>Agregar TODO</h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <input
          ref={input}
          type="text"
          name="description"
          placeholder="Aprender"
          autoComplete="off"
          className="form-control"
          onChange={handleInputChange}
          value={formValues.description}
        />
        <div className="d-grid gap-2">
          <button type="submit" className="btn btn-outline-primary mt-1">
            Agregar
          </button>
        </div>
      </form>
    </>
  );
};
