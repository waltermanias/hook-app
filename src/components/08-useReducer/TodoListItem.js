import React from "react";
import PropTypes from "prop-types";

const TodoListItem = ({ index, todo, handleToggle, handleDelete }) => {
  return (
    <li className="list-group-item" key={todo.id}>
      <p
        onClick={() => handleToggle(todo.id)}
        className={`${todo.done ? "complete" : ""}`}
      >{`${index + 1}. ${todo.desc}`}</p>
      <button className="btn btn-danger" onClick={() => handleDelete(todo.id)}>
        Borrar
      </button>
    </li>
  );
};

TodoListItem.propTypes = {
  todo: PropTypes.object.isRequired,
  handleToggle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
};

export default TodoListItem;
