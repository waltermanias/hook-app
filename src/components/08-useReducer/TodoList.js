import React from "react";
import PropTypes from "prop-types";
import TodoListItem from "./TodoListItem";

export const TodoList = ({ todos, handleToggle, handleDelete }) => {
  return (
    <ul className="list-group list-group-flush">
      {todos.map((todo, i) => {
        return (
          <TodoListItem
            handleDelete={handleDelete}
            handleToggle={handleToggle}
            todo={todo}
            key={todo.id}
            index={i}
          />
        );
      })}
    </ul>
  );
};

TodoList.propTypes = {
  todos: PropTypes.array.isRequired,
  handleToggle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
};

export default TodoList;
