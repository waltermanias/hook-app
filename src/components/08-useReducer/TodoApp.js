import React, { useReducer, useEffect } from "react";
import { todoReducer } from "./todoReducer";
import { TodoList } from "./TodoList";
import { TodoAdd } from "./TodoAdd";

import "./styles.css";

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (id) => {
    dispatch({
      type: "delete",
      payload: id,
    });
  };

  const handleToggle = (todoId) => {
    dispatch({
      type: "toggle",
      payload: todoId,
    });
  };

  const handleAddTodo = (newTodo) => {
    dispatch({
      type: "add",
      payload: newTodo,
    });
  };

  return (
    <>
      <h1>
        Todo App <small>({todos.length})</small>
      </h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            handleToggle={handleToggle}
            handleDelete={handleDelete}
          />
        </div>
        <div className="col-5">
          <TodoAdd handleAddTodo={handleAddTodo} />
        </div>
      </div>
    </>
  );
};
