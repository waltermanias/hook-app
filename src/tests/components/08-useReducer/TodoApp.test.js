import { mount } from "enzyme";
import { act } from "@testing-library/react";
import { TodoApp } from "../../../components/08-useReducer/TodoApp";

describe("TodoApp", () => {
  Storage.prototype.setItem = jest.fn(() => {});

  afterEach(() => {
    Storage.prototype.setItem.mockReset();
  });

  it("should render the component", () => {
    const wrapper = mount(<TodoApp />);

    expect(wrapper).toMatchSnapshot();
  });

  it("should render the component 2", () => {
    const wrapper = mount(<TodoApp />);

    act(() => {
      wrapper.find("TodoAdd").prop("handleAddTodo")({
        done: false,
        desc: "test-1",
        id: 1,
      });
    });

    expect(wrapper.find("h1").text().trim()).toBe("Todo App (1)");
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
  });
});
