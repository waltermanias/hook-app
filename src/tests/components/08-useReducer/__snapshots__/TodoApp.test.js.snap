// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp should render the component 1`] = `
Array [
  <h1>
    Todo App 
    <small>
      (
      0
      )
    </small>
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <ul
        className="list-group list-group-flush"
      />
    </div>
    <div
      className="col-5"
    >
      Array [
        <h4>
          Agregar TODO
        </h4>,
        <hr />,
        <form
          onSubmit={[Function]}
        >
          <input
            autoComplete="off"
            className="form-control"
            name="description"
            onChange={[Function]}
            placeholder="Aprender"
            type="text"
            value=""
          />
          <div
            className="d-grid gap-2"
          >
            <button
              className="btn btn-outline-primary mt-1"
              type="submit"
            >
              Agregar
            </button>
          </div>
        </form>,
      ]
    </div>
  </div>,
]
`;
